pkg/controllers/node/emptiness.go

```diff
@@ -0,0 +1,91 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha3"
+	"github.com/awslabs/karpenter/pkg/utils/functional"
+	"github.com/awslabs/karpenter/pkg/utils/node"
+	"github.com/awslabs/karpenter/pkg/utils/pod"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Emptiness is a subreconciler that terminates nodes that are empty after a ttl
+type Emptiness struct {
+	kubeClient client.Client
+}
+
+// Reconcile reconciles the node
+func (r *Emptiness) Reconcile(ctx context.Context, provisioner *v1alpha3.Provisioner, n *v1.Node) (reconcile.Result, error) {
+	// 1. Ignore node if not applicable
+	if provisioner.Spec.TTLSecondsAfterEmpty == nil {
+		return reconcile.Result{}, nil
+	}
+	if !node.IsReady(n) {
+		return reconcile.Result{}, nil
+	}
+	// 2. Remove ttl if not empty
+	empty, err := r.isEmpty(ctx, n)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if !empty {
+		if _, ok := n.Annotations[v1alpha3.ProvisionerTTLAfterEmptyKey]; ok {
+			delete(n.Annotations, v1alpha3.ProvisionerTTLAfterEmptyKey)
+			logging.FromContext(ctx).Infof("Removed emptiness TTL from node %s", n.Name)
+		}
+		return reconcile.Result{}, nil
+	}
+	// 3. Set TTL if not set
+	n.Annotations = functional.UnionStringMaps(n.Annotations)
+	if _, ok := n.Annotations[v1alpha3.ProvisionerTTLAfterEmptyKey]; !ok {
+		ttl := time.Duration(ptr.Int64Value(provisioner.Spec.TTLSecondsAfterEmpty)) * time.Second
+		n.Annotations[v1alpha3.ProvisionerTTLAfterEmptyKey] = time.Now().Add(ttl).Format(time.RFC3339)
+		logging.FromContext(ctx).Infof("Added TTL (+%s) to empty node %s", ttl, n.Name)
```

Comment:
I remember discussing the potential for us to change the TTL to a "TTL timestamp" indicating where it was added. We would then grab the TTL value from the provisioner to see if it's been that long.  This would help us with cases where the provisioner TTL value changes, and change up this log line. WDYT?

Reply:
I'd love to run this by @akestner, as I think it's a product decision. I'm not certain which is cleanest conceptually.

Reply:
Tentatively, I think this is the right call.